### PR TITLE
Hybrid Recurrent Checkpoint Fix

### PR DESCRIPTION
**Pull request by stale2000**
## Overview

Fix recurrent and hybrid prompt-cache checkpoint selection when `--cache-reuse` finds matching token chunks after the first prompt mismatch.

The saved longest-common-prefix boundary is now captured before chunk reuse extends `n_past`, so an exact recurrent checkpoint can never be restored beyond the actual shared prefix.

### The problem

Transformer KV cache entries are token-addressable: the runtime can often keep cache state before an edit and remove entries after it.

A recurrent layer has one accumulated state. Its state at token 20,000 contains the effect of every earlier token. If the prompt changes at token 15,000, the checkpoint at token 20,000 no longer matches the request, even when surrounding cache position metadata appears reusable.

```text
Prompt changed at token X.

Safe checkpoint:   saved at or before X.
Unsafe checkpoint: saved after X.
```

For exact recurrent state, the saved token boundary (`n_tokens`) is authoritative. Selecting or retaining checkpoints from range-like position metadata can either restore obsolete history or discard safe earlier checkpoints and force a full re-prefill.

### The fix

For exact recurrent states, the server selects a checkpoint only when:

```text
state is exact when sequence removal is FULL or bounded recurrent (RS)

safe checkpoint:
    checkpoint.n_tokens <= longest_common_prefix
    checkpoint.n_tokens < new_prompt_length
```

The second condition ensures that an identical prompt still evaluates at least one token to obtain logits.

It restores the newest checkpoint within the common prefix, replays only the suffix, and discards checkpoints beyond the common prefix. Non-exact cache handling is unchanged.

The common-prefix value is captured before optional `--cache-reuse` shifts matching chunks from after the first mismatch. Those chunks do not form part of a recurrent state shared by the two prompts.

### Benefits

The patch prevents invalid recurrent-state restoration while retaining the newest safe checkpoint. For edited or branched long prompts, avoided prefill is approximately:

```text
avoided prefill fraction = 1 - replayed_tokens / full_prompt_tokens
```

| Prompt-cache metric | Full re-prefill | Checkpoint restore | Improvement |
| --- | ---: | ---: | ---: |
| Prompt tokens evaluated | 2,811 | 517 | 81.6% fewer |
| Prompt evaluation time | 12.68 s | 2.58 s | 4.9x faster |

Example workload:

```text
32,000-token prompt
prompt changes at token 28,000
checkpoints every 1,000 tokens

newest safe checkpoint: token 27,000
replayed tokens: about 5,000
full cold prefill: 32,000 tokens

avoided work: 1 - 5,000 / 32,000 = 84.4%
```

If no checkpoint exists before the changed token, the server safely falls back to a cold prefill. The patch does not trade correctness for cache reuse.

## Requirements

- I have read and agree with the [contributing guidelines](https://github.com/ggml-org/llama.cpp/blob/master/CONTRIBUTING.md)
- AI usage disclosure: YES - Codex was used to review the cache-reuse interaction, implement this narrow fix

### CODE REVIEW

**Comment by stale2000** [comment]
> Hi @stale2000, thanks for your contribution!
> 
> Per our [contribution guidelines](https://github.com/ggml-org/llama.cpp/blob/master/CONTRIBUTING.md), the automated PR checker found the following issue(s) that need your attention:
> 
> * **PR Template not respected**: Please respect the [template](https://github.com/ggml-org/llama.cpp/blob/master/.github/pull_request_template.md?plain=1) when creating a new pull request. Make sure to fill out all required sections.
> 
> _Please note that maintainers reserve the right to make final decisions on PRs. If you believe there is a mistake, please comment below._

Issue fixed.

**Comment by ggerganov** [comment]
> The saved longest-common-prefix boundary is now captured before chunk reuse extends n_past, so an exact recurrent checkpoint can never be restored beyond the actual shared prefix.

The cache reuse functionality has nothing to do with recurrent checkpoints - it is always disabled in this case:

https://github.com/ggml-org/llama.cpp/blob/1269cb1ff1598751f846241be90083ae9ad036fb/tools/server/server-context.cpp#L1268-L1278

Provide a repro of what you are trying to fix. There is a very high probability that things are working correctly and you are trying to fix something that is not broken in the first place.
